File: previous-gen-apps/voice-kraliki/backend/app/api/compliance.py
"""
Compliance API endpoints

Provides endpoints for managing consent, retention policies,
and compliance-related operations.
"""

from datetime import UTC, datetime
from enum import Enum
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel

# Import compliance service - will be added after service integration
# from app.services.compliance import (
from app.middleware.rate_limit import (
    API_RATE_LIMIT,
    WRITE_OPERATION_RATE_LIMIT,
    limiter,
)

#     compliance_service,
#     ConsentType,
#     ConsentStatus,
#     Region
# )

# Temporary placeholder classes
class ConsentType(str, Enum):
    RECORDING = "recording"
    TRANSCRIPTION = "transcription"
    AI_PROCESSING = "ai_processing"
    DATA_STORAGE = "data_storage"
    MARKETING = "marketing"
    ANALYTICS = "analytics"

class ConsentStatus(str, Enum):
    GRANTED = "granted"
    DENIED = "denied"
    WITHDRAWN = "withdrawn"
    EXPIRED = "expired"
    PENDING = "pending"

class Region(str, Enum):
    US = "us"
    EU = "eu"
    UK = "uk"
    CA = "ca"
    AU = "au"
    APAC = "apac"

# Placeholder compliance service
class ComplianceService:
    def capture_consent(self, **kwargs):
        return "consent_123"
    def check_consent(self, **kwargs):
        return True
    def withdraw_consent(self, **kwargs):
        return True
    def get_consent_records(self, **kwargs):
        return []
    def check_retention_compliance(self, **kwargs):
        return True
    def get_retention_policy(self, region, data_type):
        return None
    def schedule_data_deletion(self, customer_phone, data_type, data_id, created_at):
        return datetime.now(UTC)
    def export_compliance_data(self, **kwargs):
        return {}
    def delete_customer_data(self, customer_phone):
        return True
    def get_compliance_events(self, **kwargs):
        return []
    def detect_region_from_phone(self, phone_number):
        return Region.US
    @property
    def retention_policies(self):
        return {}

compliance_service = ComplianceService()

# Temporary placeholder for User model
class User:
    def __init__(self, id: str, username: str):
        self.id = id
        self.username = username

async def get_current_user():
    # Temporary placeholder - in real implementation, get from auth token
    return User(id="user_123", username="test_user")

router = APIRouter(prefix="/api/compliance", tags=["compliance"])

# Request/Response models
class ConsentRequest(BaseModel):
    session_id: str
    customer_phone: str
    consent_type: ConsentType
    status: ConsentStatus
    method: str = "verbal"
    metadata: dict[str, Any] | None = None

class ConsentCheckRequest(BaseModel):
    customer_phone: str
    consent_type: ConsentType
    session_id: str | None = None

class ConsentWithdrawRequest(BaseModel):
    consent_id: str
    session_id: str | None = None

class RetentionCheckRequest(BaseModel):
    customer_phone: str
    data_type: str
    created_at: datetime

class DataExportRequest(BaseModel):
    customer_phone: str
    format: str = "json"

@router.post("/consent")
async def capture_consent(
    request: ConsentRequest,
    current_user: User = Depends(get_current_user)
):
    """Capture consent for a specific type"""
    try:
        # Get client information
        ip_address = "127.0.0.1"
        user_agent = "Unknown"

        consent_id = compliance_service.capture_consent(
            session_id=request.session_id,
            customer_phone=request.customer_phone,
            consent_type=request.consent_type,
            status=request.status,
            method=request.method,
            ip_address=ip_address,
            user_agent=user_agent,
            metadata=request.metadata
        )

        return {
            "consent_id": consent_id,
            "status": "captured",
            "timestamp": datetime.now(UTC).isoformat()
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to capture consent: {str(e)}")

@limiter.limit(WRITE_OPERATION_RATE_LIMIT)
@router.post("/consent/check")
async def check_consent(request: Request, check_data: ConsentCheckRequest):
    """Check if valid consent exists"""
    try:
        has_consent = compliance_service.check_consent(
            customer_phone=check_data.customer_phone,
            consent_type=check_data.consent_type,
            session_id=check_data.session_id
        )

        # Get region for additional context
        region = compliance_service.detect_region_from_phone(check_data.customer_phone)

        return {
            "has_consent": has_consent,
            "customer_phone": check_data.customer_phone,
            "consent_type": check_data.consent_type.value,
            "region": region.value,
            "checked_at": datetime.now(UTC).isoformat()
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to check consent: {str(e)}")

@router.post("/consent/withdraw")
async def withdraw_consent(
    request: ConsentWithdrawRequest,
    current_user: User = Depends(get_current_user)
):
    """Withdraw previously granted consent"""
    try:
        success = compliance_service.withdraw_consent(
            consent_id=request.consent_id,
            session_id=request.session_id
        )

        if not success:
            raise HTTPException(status_code=404, detail="Consent record not found")

        return {
            "consent_id": request.consent_id,
            "status": "withdrawn",
            "timestamp": datetime.now(UTC).isoformat()
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to withdraw consent: {str(e)}")

@router.get("/consent")
async def get_consent_records(
    customer_phone: str | None = None,
    session_id: str | None = None,
    consent_type: ConsentType | None = None,
    current_user: User = Depends(get_current_user)
):
    """Get consent records with optional filtering"""
    try:
        records = compliance_service.get_consent_records(
            customer_phone=customer_phone,
            session_id=session_id,
            consent_type=consent_type
        )

        # Convert to dict for JSON response
        records_data = []
        for record in records:
            record_dict = {
                "id": record.id,
                "session_id": record.session_id,
                "customer_phone": record.customer_phone,
                "region": record.region.value,
                "consent_type": record.consent_type.value,
                "status": record.status.value,
                "method": record.method,
                "granted_at": record.granted_at.isoformat() if record.granted_at else None,
                "denied_at": record.denied_at.isoformat() if record.denied_at else None,
                "withdrawn_at": record.withdrawn_at.isoformat() if record.withdrawn_at else None,
                "expires_at": record.expires_at.isoformat() if record.expires_at else None,
                "metadata": record.metadata
            }
            records_data.append(record_dict)

        return {
            "records": records_data,
            "count": len(records_data),
            "filters": {
                "customer_phone": customer_phone,
                "session_id": session_id,
                "consent_type": consent_type.value if consent_type else None
            }
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get consent records: {str(e)}")

@limiter.limit(WRITE_OPERATION_RATE_LIMIT)
@router.post("/retention/check")
async def check_retention_compliance(request: Request, retention_data: RetentionCheckRequest):
    """Check if data can still be retained based on retention policies"""
    try:
        can_retain = compliance_service.check_retention_compliance(
            customer_phone=retention_data.customer_phone,
            data_type=retention_data.data_type,
            created_at=retention_data.created_at
        )

        # Get region and policy details
        region = compliance_service.detect_region_from_phone(retention_data.customer_phone)
        policy = compliance_service.get_retention_policy(region=region, data_type=retention_data.data_type)

        response_data = {
            "can_retain": can_retain,
            "customer_phone": retention_data.customer_phone,
            "data_type": retention_data.data_type,
            "region": region.value,
            "created_at": retention_data.created_at.isoformat(),
            "checked_at": datetime.now(UTC).isoformat()
        }

        if policy:
            response_data["policy"] = {
                "retention_days": policy.retention_days,
                "auto_delete": policy.auto_delete,
                "requires_consent": policy.requires_consent,
                "anonymize_after_retention": policy.anonymize_after_retention
            }

        return response_data

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to check retention compliance: {str(e)}")

@limiter.limit(WRITE_OPERATION_RATE_LIMIT)
@router.post("/retention/schedule")
async def schedule_data_deletion(request: Request, schedule_data: RetentionCheckRequest):
    """Schedule data for deletion based on retention policies"""
    try:
        deletion_date = compliance_service.schedule_data_deletion(
            customer_phone=schedule_data.customer_phone,
            data_type=schedule_data.data_type,
            data_id="",  # Would be provided in real implementation
            created_at=schedule_data.created_at
        )

        return {
            "customer_phone": schedule_data.customer_phone,
            "data_type": schedule_data.data_type,
            "deletion_date": deletion_date.isoformat() if deletion_date else None,
            "scheduled_at": datetime.now(UTC).isoformat()
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to schedule deletion: {str(e)}")

@router.post("/export")
async def export_customer_data(
    request: DataExportRequest,
    current_user: User = Depends(get_current_user)
):
    """Export all compliance data for a customer (GDPR right to access)"""
    try:
        export_data = compliance_service.export_compliance_data(
            customer_phone=request.customer_phone,
            format=request.format
        )

        return JSONResponse(
            content=export_data,
            headers={
                "Content-Disposition": f"attachment; filename=compliance_data_{request.customer_phone}.json"
            }
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to export data: {str(e)}")

@router.delete("/customer/{customer_phone}")
async def delete_customer_data(
    customer_phone: str,
    current_user: User = Depends(get_current_user)
):
    """Delete all compliance data for a customer (GDPR right to be forgotten)"""
    try:
        success = compliance_service.delete_customer_data(customer_phone)

        if not success:
            raise HTTPException(status_code=404, detail="Customer data not found")

        return {
            "customer_phone": customer_phone,
            "status": "deleted",
            "timestamp": datetime.now(UTC).isoformat()
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete customer data: {str(e)}")

@router.get("/events")
async def get_compliance_events(
    customer_phone: str | None = None,
    session_id: str | None = None,
    event_type: str | None = None,
    limit: int = 100,
    current_user: User = Depends(get_current_user)
):
    """Get compliance events with optional filtering"""
    try:
        events = compliance_service.get_compliance_events(
            customer_phone=customer_phone,
            session_id=session_id,
            event_type=event_type,
            limit=limit
        )

        # Convert to dict for JSON response
        events_data = []
        for event in events:
            event_dict = {
                "id": event.id,
                "event_type": event.event_type,
                "session_id": event.session_id,
                "customer_phone": event.customer_phone,
                "region": event.region.value,
                "timestamp": event.timestamp.isoformat(),
                "details": event.details,
                "user_id": event.user_id
            }
            events_data.append(event_dict)

        return {
            "events": events_data,
            "count": len(events_data),
            "filters": {
                "customer_phone": customer_phone,
                "session_id": session_id,
                "event_type": event_type,
                "limit": limit
            }
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get compliance events: {str(e)}")

@router.get("/policies")
async def get_retention_policies(
    region: Region | None = None,
    current_user: User = Depends(get_current_user)
):
    """Get retention policies"""
    try:
        policies = {}

        for policy_key, policy in compliance_service.retention_policies.items():
            if region and policy.region != region:
                continue

            policies[policy_key] = {
                "region": policy.region.value,
                "data_type": policy.data_type,
                "retention_days": policy.retention_days,
                "auto_delete": policy.auto_delete,
                "requires_consent": policy.requires_consent,
                "anonymize_after_retention": policy.anonymize_after_retention
            }

        return {
            "policies": policies,
            "count": len(policies),
            "region_filter": region.value if region else None
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get retention policies: {str(e)}")

@limiter.limit(API_RATE_LIMIT)
@router.get("/region/{phone_number}")
async def detect_region(request: Request, phone_number: str):
    """Detect region from phone number"""
    try:
        region = compliance_service.detect_region_from_phone(phone_number)

        return {
            "phone_number": phone_number,
            "region": region.value,
            "detected_at": datetime.now(UTC).isoformat()
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to detect region: {str(e)}")

File: previous-gen-apps/voice-kraliki/frontend/src/lib/services/compliance.ts
import { apiDelete, apiGet, apiPost } from '$lib/utils/api';

// Compliance interfaces matching backend models
export type ConsentType = 'recording' | 'transcription' | 'ai_processing' | 'data_storage' | 'marketing' | 'analytics';
export type ConsentStatus = 'granted' | 'denied' | 'withdrawn' | 'expired' | 'pending';
export type Region = 'us' | 'eu' | 'uk' | 'ca' | 'au' | 'apac';

export interface ConsentRecord {
	id: string;
	session_id: string;
	customer_phone: string;
	region: Region;
	consent_type: ConsentType;
	status: ConsentStatus;
	method: string;
	granted_at?: string;
	denied_at?: string;
	withdrawn_at?: string;
	expires_at?: string;
	metadata?: Record<string, any>;
}

export interface RetentionPolicy {
	region: Region;
	data_type: string;
	retention_days: number;
	auto_delete: boolean;
	requires_consent: boolean;
	anonymize_after_retention: boolean;
}

export interface ComplianceEvent {
	id: string;
	event_type: string;
	session_id: string;
	customer_phone: string;
	region: Region;
	timestamp: string;
	details: Record<string, any>;
	user_id: string;
}

// Request interfaces
export interface ConsentRequest {
	session_id: string;
	customer_phone: string;
	consent_type: ConsentType;
	status: ConsentStatus;
	method?: string;
	metadata?: Record<string, any>;
}

export interface ConsentCheckRequest {
	customer_phone: string;
	consent_type: ConsentType;
	session_id?: string;
}

export interface ConsentWithdrawRequest {
	consent_id: string;
	session_id?: string;
}

export interface RetentionCheckRequest {
	customer_phone: string;
	data_type: string;
	created_at: string;
}

export interface DataExportRequest {
	customer_phone: string;
	format?: string;
}

// Response interfaces
export interface ConsentCaptureResponse {
	consent_id: string;
	status: string;
	timestamp: string;
}

export interface ConsentCheckResponse {
	has_consent: boolean;
	customer_phone: string;
	consent_type: ConsentType;
	region: Region;
	checked_at: string;
}

export interface ConsentWithdrawResponse {
	consent_id: string;
	status: string;
	timestamp: string;
}

export interface ConsentRecordsResponse {
	records: ConsentRecord[];
	count: number;
	filters: {
		customer_phone?: string;
		session_id?: string;
		consent_type?: ConsentType;
	};
}

export interface RetentionCheckResponse {
	can_retain: boolean;
	customer_phone: string;
	data_type: string;
	region: Region;
	created_at: string;
	checked_at: string;
	policy?: {
		retention_days: number;
		auto_delete: boolean;
		requires_consent: boolean;
		anonymize_after_retention: boolean;
	};
}

export interface RetentionScheduleResponse {
	customer_phone: string;
	data_type: string;
	deletion_date?: string;
	scheduled_at: string;
}

export interface CustomerDataDeleteResponse {
	customer_phone: string;
	status: string;
	timestamp: string;
}

export interface ComplianceEventsResponse {
	events: ComplianceEvent[];
	count: number;
	filters: {
		customer_phone?: string;
		session_id?: string;
		event_type?: string;
		limit: number;
	};
}

export interface RetentionPoliciesResponse {
	policies: Record<string, {
		region: Region;
		data_type: string;
		retention_days: number;
		auto_delete: boolean;
		requires_consent: boolean;
		anonymize_after_retention: boolean;
	}>;
	count: number;
	region_filter?: Region;
}

export interface RegionDetectionResponse {
	phone_number: string;
	region: Region;
	detected_at: string;
}

// Consent Management Operations
export function captureConsent(request: ConsentRequest) {
	return apiPost<ConsentCaptureResponse>('/api/compliance/consent', request);
}

export function checkConsent(request: ConsentCheckRequest) {
	return apiPost<ConsentCheckResponse>('/api/compliance/consent/check', request);
}

export function withdrawConsent(request: ConsentWithdrawRequest) {
	return apiPost<ConsentWithdrawResponse>('/api/compliance/consent/withdraw', request);
}

export function getConsentRecords(
	customerPhone?: string,
	sessionId?: string,
	consentType?: ConsentType
) {
	const params = new URLSearchParams();
	if (customerPhone) params.set('customer_phone', customerPhone);
	if (sessionId) params.set('session_id', sessionId);
	if (consentType) params.set('consent_type', consentType);
	
	const query = params.toString();
	return apiGet<ConsentRecordsResponse>(`/api/compliance/consent${query ? `?${query}` : ''}`);
}

// Retention Management Operations
export function checkRetentionCompliance(request: RetentionCheckRequest) {
	return apiPost<RetentionCheckResponse>('/api/compliance/retention/check', request);
}

export function scheduleDataDeletion(request: RetentionCheckRequest) {
	return apiPost<RetentionScheduleResponse>('/api/compliance/retention/schedule', request);
}

export function getRetentionPolicies(region?: Region) {
	const params = new URLSearchParams();
	if (region) params.set('region', region);
	
	const query = params.toString();
	return apiGet<RetentionPoliciesResponse>(`/api/compliance/policies${query ? `?${query}` : ''}`);
}

// Data Rights Operations (GDPR)
export function exportCustomerData(request: DataExportRequest) {
	return apiPost<Record<string, any>>('/api/compliance/export', request);
}

export function deleteCustomerData(customerPhone: string) {
	return apiDelete<CustomerDataDeleteResponse>(`/api/compliance/customer/${customerPhone}`);
}

// Compliance Events
export function getComplianceEvents(
	customerPhone?: string,
	sessionId?: string,
	eventType?: string,
	limit = 100
) {
	const params = new URLSearchParams({
		limit: limit.toString()
	});
	if (customerPhone) params.set('customer_phone', customerPhone);
	if (sessionId) params.set('session_id', sessionId);
	if (eventType) params.set('event_type', eventType);
	
	const query = params.toString();
	return apiGet<ComplianceEventsResponse>(`/api/compliance/events${query ? `?${query}` : ''}`);
}

// Region Detection
export function detectRegion(phoneNumber: string) {
	return apiGet<RegionDetectionResponse>(`/api/compliance/region/${phoneNumber}`);
}

// Utility functions for common compliance operations
export interface ConsentCheckBatch {
	customer_phone: string;
	consent_type: ConsentType;
	session_id?: string;
}

export function batchConsentCheck(requests: ConsentCheckBatch[]) {
	const promises = requests.map(request => checkConsent(request));
	return Promise.all(promises);
}

export function captureMultipleConsents(
	sessionId: string,
	customerPhone: string,
	consentTypes: ConsentType[],
	status: ConsentStatus = 'granted'
) {
	const promises = consentTypes.map(consentType =>
		captureConsent({
			session_id: sessionId,
			customer_phone: customerPhone,
			consent_type: consentType,
			status
		})
	);
	return Promise.all(promises);
}

export function getCustomerConsentProfile(customerPhone: string) {
	return getConsentRecords(customerPhone).then(response => {
		const consents = response.records;
		const profile: Record<ConsentType, ConsentRecord[]> = {
			recording: [],
			transcription: [],
			ai_processing: [],
			data_storage: [],
			marketing: [],
			analytics: []
		};
		
		consents.forEach(consent => {
			if (profile[consent.consent_type]) {
				profile[consent.consent_type].push(consent);
			}
		});
		
		return {
			customer_phone: customerPhone,
			total_consents: consents.length,
			active_consents: consents.filter(c => c.status === 'granted').length,
			consents_by_type: profile,
			region: consents[0]?.region || 'us'
		};
	});
}

export function checkDataRetentionForCustomer(
	customerPhone: string,
	dataTypes: string[],
	createdAt: string
) {
	const promises = dataTypes.map(dataType =>
		checkRetentionCompliance({
			customer_phone: customerPhone,
			data_type: dataType,
			created_at: createdAt
		})
	);
	
	return Promise.all(promises).then(responses => 
		responses.reduce((acc, response, index) => {
			acc[dataTypes[index]] = response;
			return acc;
		}, {} as Record<string, RetentionCheckResponse>)
	);
}

// Compliance monitoring utilities
export function createComplianceMonitor(refreshInterval = 60000) {
	let intervalId: ReturnType<typeof setInterval> | null = null;
	
	const startMonitoring = (callback: (events: ComplianceEventsResponse) => void) => {
		const fetchEvents = async () => {
			try {
				const events = await getComplianceEvents(undefined, undefined, undefined, 50);
				callback(events);
			} catch (error) {
				console.error('Failed to fetch compliance events:', error);
			}
		};
		
		// Initial fetch
		fetchEvents();
		
		// Set up interval
		intervalId = setInterval(fetchEvents, refreshInterval);
	};
	
	const stopMonitoring = () => {
		if (intervalId) {
			clearInterval(intervalId);
			intervalId = null;
		}
	};
	
	return {
		startMonitoring,
		stopMonitoring
	};
}

// Consent workflow helpers
export interface ConsentWorkflowConfig {
	require_recording: boolean;
	require_transcription: boolean;
	require_ai_processing: boolean;
	require_data_storage: boolean;
	auto_capture: boolean;
	expiry_days?: number;
}

export function executeConsentWorkflow(
	sessionId: string,
	customerPhone: string,
	config: ConsentWorkflowConfig,
	status: ConsentStatus = 'granted'
) {
	const consentTypes: ConsentType[] = [];
	
	if (config.require_recording) consentTypes.push('recording');
	if (config.require_transcription) consentTypes.push('transcription');
	if (config.require_ai_processing) consentTypes.push('ai_processing');
	if (config.require_data_storage) consentTypes.push('data_storage');
	
	if (consentTypes.length === 0) {
		return Promise.resolve([]);
	}
	
	return captureMultipleConsents(sessionId, customerPhone, consentTypes, status);
}